go/src/meguca/geoip/db.go

```diff
@@ -4,38 +4,94 @@ import (
 	"net"
 	"strings"
 	"time"
+	"os"
+	"errors"
+	"fmt"
+	"sync"
+	"io/ioutil"
+	"net/http"
+	"crypto/md5"
+	"path/filepath"
+
+	"meguca/db"
 
 	"github.com/oschwald/maxminddb-golang"
 	"github.com/go-playground/log"
+	"github.com/mholt/archiver"
 )
 
 // nil, if database not loaded
-var db *maxminddb.Reader
+var (
+	// Ensure we only load the old GeoIP DB once on server start
+	once sync.Once
+	// Ensures no data races
+	rw sync.RWMutex
+	// GeoIP database
+	gdb *maxminddb.Reader
+)
 
+// NY location
 var NY *time.Location
 
 func init() {
 	NY, _ = time.LoadLocation("America/New_York")
 }
 
+// Load checks if the GeoLite DB exists, and calls load it if it does
 func Load() error {
-	var err error
-	db, err = maxminddb.Open("GeoLite2-Country.mmdb")
-	if err != nil {
-		db = nil
+	if err := check(); err != nil {
+		rw.Lock()
+		gdb = nil
+		rw.Unlock()
+		log.Warn("Unable to use GeoLite DB: ", err)
 	}
+
 	return nil
 }
 
-// Look up the country ISO code of the IP
+// loads the GeoLite DB
+func load() (err error) {
+	rw.Lock()
+	defer rw.Unlock()
+	gdb, err = maxminddb.Open("GeoLite2-Country.mmdb")
+	return
+}
+
+// reloads the GeoLite DB
+func reload() error {
+	rw.Lock()
+
```

Comment:
defer is your friend

Reply:
I used explicit unlocking for performance reasons.